hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java

```diff
@@ -70,29 +73,37 @@ public boolean test(final FilterableMessage message) {
 
     private List<Object> select(Object record, ListIterator<String> iter) {
         Object current = record;
-        while (iter.hasNext() && current instanceof GenericRecord) {
-            GenericRecord currentRecord = (GenericRecord) current;
-            String selector = iter.next();
-            Matcher arrayMatcher = ARRAY_PATTERN.matcher(selector);
+        while (iter.hasNext()) {
+            if (current instanceof GenericRecord) {
+                GenericRecord currentRecord = (GenericRecord) current;
+                String selector = iter.next();
+                Matcher arrayMatcher = ARRAY_PATTERN.matcher(selector);
+
+                if (arrayMatcher.matches()) {
+                    String idx = arrayMatcher.group(GROUP_IDX);
+                    selector = arrayMatcher.group(GROUP_SELECTOR);
+
+                    current = currentRecord.get(selector);
+                    if (! (current instanceof GenericArray)) {
+                        return emptyList();
+                    }
+
+                    GenericArray<Object> currentArray = (GenericArray) current;
+                    if (idx.equals(WILDCARD_IDX)) {
+                        return selectMultipleArrayItems(iter, currentArray);
+                    } else {
+                        current = selectSingleArrayItem(Integer.valueOf(idx), currentArray);
+                    }
 
-            if (arrayMatcher.matches()) {
-                String idx = arrayMatcher.group(GROUP_IDX);
-                selector = arrayMatcher.group(GROUP_SELECTOR);
-
-                current = currentRecord.get(selector);
-                if (! (current instanceof GenericArray)) {
-                    return emptyList();
-                }
-
-                GenericArray<Object> currentArray = (GenericArray) current;
-                if (idx.equals(WILDCARD_IDX)) {
-                    return selectMultipleArrayItems(iter, currentArray);
                 } else {
-                    current = selectSingleArrayItem(Integer.valueOf(idx), currentArray);
+                    current = currentRecord.get(selector);
                 }
-
-            } else {
+            } else if (current instanceof HashMap) {
+                Map<Utf8, Object> currentRecord = (HashMap<Utf8, Object>) current;
+                Utf8 selector = new Utf8(iter.next());
                 current = currentRecord.get(selector);
+            } else {
+                break;
```

Comment:
Note - this break is substitute for `while (iter.hasNext() && (current instanceof GenericRecord) || current instanceof HashMap))`. I racked my brain for a bit to get into that. What do you think about getting rid of this `else { break; }` and to use this `iter.hasNext() && (current instanceof GenericRecord) || current instanceof HashMap)` boolean expression (and extracting this boolean expression to separate function)?

Reply:
I add new method for decide if analyzed record type is supported